connectors/agentic-ai/src/main/java/io/camunda/connector/agenticai/adhoctoolsschema/resolver/CamundaClientAdHocToolsSchemaResolver.java

```diff
@@ -0,0 +1,134 @@
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+import org.camunda.bpm.model.xml.instance.ModelElementInstance;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CamundaClientAdHocToolsSchemaResolver implements AdHocToolsSchemaResolver {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(CamundaClientAdHocToolsSchemaResolver.class);
+
+  private static final String INPUT_SCHEMA_PROPERTY_NAME = "camunda:adHocActivityInputSchema";
+
+  private final CamundaClient camundaClient;
+  private final ObjectMapper objectMapper;
+  private final FeelEngineWrapper feelEngineWrapper;
+
+  public CamundaClientAdHocToolsSchemaResolver(
+      CamundaClient camundaClient, ObjectMapper objectMapper, FeelEngineWrapper feelEngineWrapper) {
+    this.camundaClient = camundaClient;
+    this.objectMapper = objectMapper;
+    this.feelEngineWrapper = feelEngineWrapper;
+  }
+
+  @Override
+  public AdHocToolsSchemaResponse resolveSchema(
+      Long processDefinitionKey, String adHocSubprocessId) {
+    LOGGER.info(
+        "Resolving tool schema for ad-hoc subprocess {} in process definition with key {}",
+        adHocSubprocessId,
+        processDefinitionKey);
+
+    final String processDefinitionXml =
+        camundaClient.newProcessDefinitionGetXmlRequest(processDefinitionKey).send().join();
+
+    final BpmnModelInstance modelInstance =
+        Bpmn.readModelFromStream(
+            new ByteArrayInputStream(processDefinitionXml.getBytes(StandardCharsets.UTF_8)));
+
+    final var processElement = modelInstance.getModelElementById(adHocSubprocessId);
+    if (!(processElement instanceof final AdHocSubProcess adHocSubProcess)) {
+      throw new RuntimeException("Ad-hoc subprocess %s not found".formatted(adHocSubprocessId));
+    }
+
+    final var toolDefinitions =
+        adHocSubProcess.getChildElementsByType(FlowNode.class).stream()
+            .filter(element -> element.getIncoming().isEmpty())
+            .map(this::mapActivityToToolDefinition)
+            .toList();
+
+    return new AdHocToolsSchemaResponse(toolDefinitions);
+  }
+
+  private AdHocToolDefinition mapActivityToToolDefinition(FlowNode element) {
+    final var documentation = getDocumentation(element);
+    final var inputSchema = getInputSchema(element);
+
+    return inputSchema
+        .map(
+            inputSchemaValue ->
+                this.toolDefinitionFromInputSchema(element, documentation, inputSchemaValue))
+        .orElseGet(
+            () -> new AdHocToolDefinition(element.getId(), documentation, JsonSchema.empty()));
+  }
+
+  private AdHocToolDefinition toolDefinitionFromInputSchema(
+      final FlowNode element, final String documentation, final String inputSchema) {
+
+    var inputSchemaJson = inputSchema;
+    if (inputSchemaJson.startsWith("=")) {
+      inputSchemaJson = feelEngineWrapper.evaluateToJson(inputSchemaJson);
```

Comment:
Basic FEEL support without external variables allows to construct the schema via an expression (e.g. looping over a list).

Reply:
Removed this when moving to the documentation block.